framework/src/org/checkerframework/common/signatureinference/SignatureInferenceScenes.java

```diff
@@ -0,0 +1,588 @@
+                            + " Exception message: " + e.getMessage());
+                }
+            }
+            scenes.put(jaifPath, scene);
+        }
+        return scenes.get(jaifPath);
+    }
+
+    /**
+     * Clears the set of modified scenes.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    public static void clearModifiedScenes() {
+        modifiedScenes.clear();
+    }
+
+    /**
+     * Adds the identifier of a Scene in the set of modified scenes.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    private static void addModifiedScene(String scene) {
+        modifiedScenes.add(scene);
+    }
+
+    /**
+     * Write all modified scenes into .jaif files.
+     * (Scenes are modified by the method
+     * {@link SignatureInferenceScenes#updateAnnotationSetInScene updateAnnotationSetInScene)}
+     */
+    public static void writeScenesToJaif() {
+        for (String jaifPath : modifiedScenes) {
+            try {
+                File jaifFile = new File(jaifPath);
+                if (!jaifFile.exists()) {
+                    jaifFile.getParentFile().mkdirs();
+                }
+                AScene scene = scenes.get(jaifPath);
+                IndexFileWriter.write(scene, new FileWriter(jaifPath));
+            } catch (IOException e) {
+                ErrorReporter.errorAbort("Could not open file in: " + jaifPath
+                        + ". Exception message: " + e.getMessage());
+            } catch (DefException e) {
+                ErrorReporter.errorAbort(e.getMessage());
+            } catch (Exception e) {
+                System.out.println();
+            }
+        }
+    }
+
+    /**
+     * Updates the parameters' types of the method methodElt in the Scene of the
+     * class with symbol classSymbol.
+     *
+     * For each method parameter in methodElt:
+     *   If the Scene does not contain an annotated type for that parameter,
+     *   then the type of the respective value passed as argument in the
+     *   method call methodInvNode will be added to the parameter in the Scene.
+     *   If the Scene previously contained an annotated type for that parameter,
+     *   then its new type will be the LUB between the previous type and the
+     *   type of the respective value passed as argument in the method call.
+     *
+     * @param methodInvNode the node representing a method invocation.
+     * @param classSymbol the symbol of the class that contains the method being
+     * invoked.
+     * @param methodElt the element of the method being invoked.
+     * @param atf the annotated type factory of a given type system, whose
+     * type hierarchy will be used to update the method parameters' types.
+     */
+    public static void updateMethodParameterTypeInScene(
```

Comment:
The "Parameters" should also be plural, right? I changed to updatedInferredMethodParametersType.